unit/goto-symex/goto_symex_state.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************************************************\
+
+Module: Unit tests for goto_symex_statet
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <testing-utils/message.h>
+#include <testing-utils/use_catch.h>
+
+#include <analyses/dirty.h>
+#include <goto-symex/goto_symex_state.h>
+#include <util/arith_tools.h>
+#include <util/namespace.h>
+#include <util/symbol_table.h>
+
+static void add_to_symbol_table(
+  symbol_tablet &symbol_table,
+  const symbol_exprt &symbol_expr)
+{
+  symbolt symbol;
+  symbol.name = symbol_expr.get_identifier();
+  symbol.type = symbol_expr.type();
+  symbol.value = symbol_expr;
+  symbol.is_thread_local = true;
+  symbol_table.insert(symbol);
+}
+
+SCENARIO(
+  "Goto symex state assignment",
+  "[core][goto-symex][goto-symex-state][assignment]")
+{
+  // Initialize goto state
+  std::list<goto_programt::instructiont> target;
+  symex_targett::sourcet source{"fun", target.begin()};
+  guard_managert manager;
+  std::size_t fresh_name_count = 1;
```

Comment:
Static variable inside the lambda is usual

Reply:
Yes but when ran in catch that makes its result a bit unpredictable because the same counter will be used in each test but we don't have a guarantee on the order, so I wouldn't be able to use assertions like `result2.get().get_identifier() == "foo!0#2"`